faiss_vector_posting.go

```diff
@@ -387,13 +387,93 @@ func (sb *SegmentBase) InterpretVectorIndex(field string, requiresFiltering bool
+					// the cluster.
+					clusterAssignment, _ := vecIndex.GetClusterAssignment()
+					// Accounting for a flat index
+					if len(clusterAssignment) == 0 {
+						scores, ids, err := vecIndex.SearchWithIDs(qVector, k,
+							vectorIDsToInclude, params)
+						if err != nil {
+							return nil, err
+						}
+
+						addIDsToPostingsList(rv, ids, scores)
+						return rv, nil
+					}
+
+					// Converting to roaring bitmap for ease of intersect ops with
+					// the set of eligible doc IDs.
+					centroidVecIDMap := make(map[int64]*roaring.Bitmap)
+					for centroidID, vecIDs := range clusterAssignment {
+						if _, exists := centroidVecIDMap[centroidID]; !exists {
+							centroidVecIDMap[centroidID] = roaring.NewBitmap()
+						}
+						for _, vecID := range vecIDs {
+							centroidVecIDMap[centroidID].Add(uint32(vecID))
+						}
+					}
+
+					// Getting the vector IDs corresponding to the eligible
+					// doc IDs.
+					eligibleVecIDsBitmap := roaring.NewBitmap()
+					for _, eligibleDocID := range eligibleDocIDs {
+						vecIDs := docVecIDMap[uint32(eligibleDocID)]
+						for _, vecID := range vecIDs {
+							eligibleVecIDsBitmap.Add(uint32(vecID))
+						}
+					}
+
+					// Determining which clusters, identified by centroid ID,
+					// have at least one eligible vector and hence, ought to be
+					// probed.
+					eligibleCentroidIDs := make([]int64, 0)
+					for centroidID, vecIDs := range centroidVecIDMap {
+						vecIDs.And(eligibleVecIDsBitmap)
+						if vecIDs.GetCardinality() > 0 {
+							// The mapping is now reduced to those vectors which
+							// are also eligible docs for the filter query.
+							centroidVecIDMap[centroidID] = vecIDs
+							eligibleCentroidIDs = append(eligibleCentroidIDs, centroidID)
+						} else {
+							// don't consider clusters with no eligible IDs.
+							delete(centroidVecIDMap, centroidID)
+						}
+					}
+
+					// Ordering the retrieved centroid IDs by increasing order
+					// of distance i.e. decreasing order of proximity to query vector.
+					closestCentroidIDs, centroidDistances, _ :=
+						vecIndex.GetCentroidDistances(qVector, eligibleCentroidIDs)
+
+					// Getting the nprobe value set at index time.
+					nprobe := vecIndex.GetNProbe()
+
+					eligibleDocsTillNow := int64(0)
+					minEligibleCentroids := 0
+					for i, centroidID := range closestCentroidIDs {
+						eligibleDocsTillNow += int64(centroidVecIDMap[centroidID].GetCardinality())
+						if eligibleDocsTillNow >= k && i >= int(nprobe-1) {
+							// Continue till at least 'K' cumulative vectors are
+							// collected or 'nprobe' clusters are examined, whichever
+							// comes later.
+							minEligibleCentroids = i + 1
+							break
+						}
+						minEligibleCentroids = i + 1
+					}
+
+					// Search the clusters specified by 'closestCentroidIDs' for
+					// vectors whose IDs are present in 'vectorIDsToInclude'
+					scores, ids, err := vecIndex.SearchSpecifiedClusters(vectorIDsToInclude,
```

Comment:
I suppose this needs updating per changes in your go-faiss PR.

Reply:
yeah, will do it after I resolve the local FAISS issues I'm facing(like we discussed yesterday).